### ISSUE TITLE
Several issues with array field for emails

### ISSUE THREAD

**Issue by woylie** [ory/kratos#2374]
Following the [documentation](https://www.ory.sh/docs/kratos/concepts/credentials/username-email-password#use-case-multiple-e-mails-and-password), I set up an identity schema that allows a user to have multiple email addresses.

```json
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "title": "Person",
  "type": "object",
  "properties": {
    "traits": {
      "type": "object",
      "properties": {
        "emails": {
          "type": "array",
          "items": {
            "type": "string",
            "format": "email",
            "ory.sh/kratos": {
              "credentials": {
                "password": {
                  "identifier": true
                },
                "totp": {
                  "account_name": true
                }
              }
            }
          }
        }
      },
      "required": ["emails"],
      "additionalProperties": false
    }
  }
}
```

Now I have a problem. When I submit the registration form with an invalid email address and then get the flow again after the redirect, Kratos returns two nodes for emails. The first one has the name `traits.emails.0` and includes the validation message, while the second one has the name `traits.emails` without a validation message (same as in a fresh registration flow before submitting). Both fields include the `required` attribute.

```elixir
%{
  "expires_at" => "2022-04-04T05:29:11.572512Z",
  "id" => "64f076f1-3754-4837-86a0-c11aa038e83e",
  "issued_at" => "2022-04-04T05:19:11.572512Z",
  "request_url" => "http://127.0.0.1:4433/self-service/registration/browser",
  "type" => "browser",
  "ui" => %{
    "action" => "http://127.0.0.1:4433/self-service/registration?flow=64f076f1-3754-4837-86a0-c11aa038e83e",
    "method" => "POST",
    "nodes" => [
      %{
        "attributes" => %{
          "disabled" => false,
          "name" => "traits.emails.0",
          "node_type" => "input",
          "type" => "text"
        },
        "group" => "default",
        "messages" => [
          %{
            "id" => 4000001,
            "text" => "\"uiae\" is not valid \"email\"",
            "type" => "error"
          }
        ],
        "meta" => %{},
        "type" => "input"
      },
      %{
        "attributes" => %{
          "disabled" => false,
          "name" => "csrf_token",
          "node_type" => "input",
          "required" => true,
          "type" => "hidden",
          "value" => "W9LR3TreQhEUuhfXpwH6M46Dp3+f/SCHLi4b3vY3lx7JxOihF+fY9QZCwxcBX/D5l2c8uhDovmDuagfGpOlwzg=="
        },
        "group" => "default",
        "messages" => [],
        "meta" => %{},
        "type" => "input"
      },
      %{
        "attributes" => %{
          "disabled" => false,
          "name" => "traits.emails",
          "node_type" => "input",
          "required" => true,
          "type" => "text"
        },
        "group" => "password",
        "messages" => [],
        "meta" => %{},
        "type" => "input"
      },
      %{
        "attributes" => %{
          "disabled" => false,
          "name" => "password",
          "node_type" => "input",
          "required" => true,
          "type" => "password"
        },
        "group" => "password",
        "messages" => [],
        "meta" => %{
          "label" => %{"id" => 1070001, "text" => "Password", "type" => "info"}
        },
        "type" => "input"
      },
      %{
        "attributes" => %{
          "disabled" => false,
          "name" => "method",
          "node_type" => "input",
          "type" => "submit",
          "value" => "password"
        },
        "group" => "password",
        "messages" => [],
        "meta" => %{
          "label" => %{
            "context" => %{},
            "id" => 1040001,
            "text" => "Sign up",
            "type" => "info"
          }
        },
        "type" => "input"
      }
    ]
  }
}
```

<img width="583" alt="image" src="https://user-images.githubusercontent.com/13847569/161647017-76479391-5537-41ed-81b9-1bb51122ccc5.png">

Did I do something wrong in the configuration?

_Originally posted by @woylie in https://github.com/ory/kratos/discussions/2359_

**Comment by woylie** [comment]
Regarding the last point about the account recovery, I didn't have the field configured for recovery. The other issues are still there.

**Comment by aeneasr** [comment]
Thank you for the report, we have no tests set up for this yet so it appears that there are several issues related to primary identifiers encapsuled in an array. This is probably not a very simple issue to tackle, but contributions are always welcomed :)

**Comment by woylie** [comment]
It seems like this is not only an issue with identifiers within arrays.

When I set up an ordinary array field like this:

```json
{
  "secondary_emails": {
    "type": "array",
    "title": "Secondary E-Mails",
    "items": {
      "type": "string",
      "format": "email"
    }
  }
}
```

I'm getting the same:

1. text input, name: `traits.secondary_emails`, value: null
2. one text input per existing secondary email, name: `traits.secondary_emails.0` etc.

The form cannot be submitted without entering an email into the empty field, because `""` is not a valid email. I modified the UI to not render the `traits.secondary_emails` field. Now I can submit the form (although I cannot add a new value, obviously). However, if the identity already has values in the `secondary_emails` field (added via the API), although inputs for all entries are rendered, submitting the form unchanged will just remove all the values. So it seems like the submit-settings endpoint does not parse those array fields correctly.

I tried to modify the UI to change the names from `traits.secondary_emails.0`, `traits.secondary_emails.1` etc. to `traits.secondary_emails[]`, which is a common way for representing arrays in HTML forms, but this is also not parsed.

**Comment by woylie** [comment]
I also tried mapping the name from `traits.secondary_emails.0` to `traits.secondary_emails[0]`, but that still wipes out all values on submit

**Comment by jonas-jonas** [comment]
@aeneasr yes, I already thought so. Will take a look.

**Comment by jonas-jonas** [comment]
I played around with the relevant code here a bit and would like a bit of feedback before I start to implement anything.

This is a bit of weird situation, because the UI "rendering" is tied closely to HTML Form fields, which don't really have a way of expressing array fields. Of course there is the option of using `traits.emails.0` which is the desired behavior here. Unfortunately, the validator only knows about `traits.emails` and doesn't match to the indexed field (`traits.emails` != `traits.emails.0`). 

The underlying issue is [here](https://github.com/ory/kratos/blob/master/ui/container/container.go#L251). It tries to find a UI node with the ID `traits.emails.0` because the validator returned an error with that `InstancePtr`. This obviously fails, the schema only mentions `traits.emails`. So it adds a UI node with ID `traits.emails.0` and attaches the error message. This is what causes the second input. 

Note: the validation error comes from the library (jsonschema) and does not mention the original field `traits.email`.

I am not really sure how to proceed here. I could change the `Find` method to match any IDs that end with `.0` (or other numbers), but that seems hacky? (What about fields that a user named `0` on purpose?) But I think that might be the best solution here. What do you think?

**Comment by woylie** [comment]
I think a more common way to represent arrays in HTML forms is the use of square brackets in input names:

```html
<form>
  <input name="emails[]" value="a@a.a" />
  <input name="emails[]" value="b@b.b" />
</form>
```

Which would result in:

```json
{
  "emails[]": "a@a.a",
  "emails[]": "b@b.b"
}
```

Which then should be parsed into:

```json
{"emails": ["a@a.a", "b@b.b"]}
```

Objects can also be represented with square brackets:

```html
<form>
  <input name="name[family]" value="LuPone" />
  <input name="name[given]" value="Patti" />
  <input name="preferences[appearance][theme]" value="dark" />
</form>
```

```json
{
  "name": {
    "family": "LuPone",
    "given": "Patti"
  },
  "preferences": {
    "appearance": {
      "theme": "dark"
    }
  }
}
```

Since with empty square brackets alone it would not be possible to represent a list of objects, integer indexes are used.

```html
<form>
  <input name="pets[0][name]" value="Benny" />
  <input name="pets[0][species]" value="dog" />
  <input name="pets[1][name]" value="Lucy" />
  <input name="pets[1][species]" value="cat" />
</form>
```

```json
{
  "pets": [
    {"name": "Benny", "species": "dog"},
    {"name": "Lucy", "species": "cat"}
  ]
}
```

I think this is a clear and unambiguous (and established) representation of arrays and objects with a very low collision risk with user-given trait names. To make this work, we'd need to:

1. Update the UI builder to generate input names as described.
2. Add a parsing step to turn names with square brackets into arrays and objects _before_ validating against the schema.

**Comment by jonas-jonas** [comment]
Thanks for the write-up. You are right, I focused on the JSON endpoint for submitting the flow and didn't actually think about actual HTML form submission as another use-case. I will investigate that a bit more. Thanks!

**Comment by aeneasr** [comment]
Thank you both for the details! As far as I remember, bracket-based encoding is something PHP introduced. For Ory we decided to use JSON Pointers (or dot notation when JSON pointers are not feasable). I want to stick with that, because introducing another method (square brackets) won't make things easier to understand. Generally, I think we want to move completely to JSON Pointers at some point but there's still lots of work that needs to be done before we can do that.

In the end, I think it's just application logic and not neccessarily how we encode the names!

**Comment by woylie** [comment]
Did you by any chance have another look into what would be involved to fix this? We're handling the array field updates outside of the settings flow via the admin API, but whenever the user submits a settings flow, the field will be wiped clean. I saw that you added a `PATCH` operation for updating identities, but it doesn't look like the settings flow can submitted with a `PATCH`, so there doesn't seem to be a workaround for this.

**Comment by aeneasr** [comment]
We unfortunately still did not have the capacity to work on this. We do know that it's an issue but there are currently many other priorities we have to deal with :(

**Comment by DylanBulmer** [comment]
Hi all! I was curious if any time has been put towards this issue. I have a common scenario where users of my application have multiple emails. Instead of having the user create multiple user identities (one per email), I allowed the identity schema to accept multiple email addresses as documented here: https://www.ory.sh/docs/kratos/manage-identities/customize-identity-schema#multiple-emails-and-password. This works, as in the email is stored in an array and I can sign in, but I cannot add additional emails to the identity via the settings UI. 

I tried multiple avenues within a custom UI such as structuring my form to submit the emails like this:
```ts
FormData: {
  "traits.emails.0": "me+email1@example.com"
  "traits.emails.1": "me+email2@example.com"
  "traits.emails.2": "me+email3@example.com"
}
```
And then I attempted this:
```ts
FormData: {
  "traits.emails": "me+email1@example.com"
  "traits.emails": "me+email2@example.com"
  "traits.emails": "me+email3@example.com"
}
```

I found that both `traits.emails` and `traits.emails.0` had to be identical in order to send successfully in the first example, and then in the second example, only the final email in the list took, but it replaced my currently verified email.

Is there a known workaround for this? Or are we stuck with only having one email per user?

I'm currently in a POC phase. The organization whom I'm assisting is trying to migrate off MongoDB's Realm/app services as Mongo has recently deprecated these and soon to shut them down. Not having this feature is not a deal breaker as I know Auth0 cannot do this either (though, they do have a clunky/hacky linking feature), having this feature would provide a better QoL for the end users.

Thanks in advance!

**Comment by aeneasr** [comment]
Hello! The use case still makes sense and yes this is still unsolved. In most cases, it's not needed to have an email array as far as we have seen, which is why it's not supported. It is definitely a good amount of work to support array emails especially if they're supposed to be login identifiers etc.

It's currently not planned/on the roadmap

**Comment by DylanBulmer** [comment]
Thanks for your response, @aeneasr. I can see it being a massive undertaking, I hope you consider it in the future!

**Comment by fboerman** [comment]
hi all, I also run into this. I understand its not a priority for the project, but is there any workaround for this known?

**Comment by dmakarenko** [comment]
@aeneasr mentioned that
> In most cases, it's not needed to have an email array as far as we have seen, which is why it's not supported 

Just would like to reiterate on what others mentioned above a few years back that this issue seems to be applicable to any optional `traits` properties defined in jsonschema as **arrays of objects**.
